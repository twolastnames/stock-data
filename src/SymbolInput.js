import React from 'react';
import ReactDom from 'react-dom'
import './App.css';

function getDisplayString({name, symbol}) {
  return `${name} (${symbol})`
}

function DataSet(props) {
  const options = props.suggestions.map((text, key) => {
    return <option key={key} value={text} />
  })
  return (
    <datalist id="suggestions">
      {options}
    </datalist>
  )
}

function SymbolInput(props) {
  let currentSuggestions = []
  let lastValue = ''
  let inFlight = false
  const produceSymbol = async () => {
    if(inFlight) {
      return
    }
    const value = document.querySelector('[name="symbol"]').value.trim()
    if(value === lastValue) {
      return
    }
    lastValue = value
    if(value.length < 3) {
      currentSuggestions = []
      ReactDom.render(<DataSet suggestions={currentSuggestions} />, document.getElementById('suggestion-container'))
      return
    }
    const match = value.match(/^(.+) \((.+)\)$/)
    if(!match || !currentSuggestions.includes(value)) {
      try {
        inFlight = true
        const response = await fetch(`/symbol/${value}`)
        const similarMatches = await response.json()
        currentSuggestions = similarMatches.map(getDisplayString)
        await ReactDom.render(
          <DataSet suggestions={currentSuggestions} />,
          document.getElementById('suggestion-container'),
          () => inFlight = false
        )
        return
      } catch {
        inFlight = false
      }
    }
    const symbol = match[2]
    const name = match[1]
    
    document.querySelector('[name="symbol"]').value = ''
    lastValue = ''
    props.symbolListener(symbol, name)
  }

  setInterval(produceSymbol, 500)

  return (
    <div>
      <div id="suggestion-container">
      <datalist id="suggestions" />
      </div>
      <label>Company Name</label>
      <input type="text" list="suggestions" name="symbol" autoComplete="off"/>
    </div>
  )
}

export default SymbolInput

