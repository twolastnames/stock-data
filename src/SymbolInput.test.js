import React from 'react';
import SymbolInput from './SymbolInput';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

jest.useFakeTimers();

test('renders blank state', () => {
  const tree = renderer.create(<SymbolInput />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('symbol completion', () => {
  let wrapper = null;
  let symbolListener = null;
  let errorListener = null;
  beforeEach(() => {
    symbolListener = jest.fn();
    errorListener = jest.fn();
    wrapper = shallow(
      <SymbolInput
        symbolListener={symbolListener}
        errorListener={errorListener}
      />
    );
    fetchMock.mockIf(/.*symbol\/john/, async () => {
      return {
        body: JSON.stringify([{ name: 'John', symbol: 'Dough' }]),
      };
    });
  });

  it('suggests symbols', async (done) => {
    const input = await wrapper.find('.name-input');
    input.simulate('change', { target: { value: 'john' } });
    jest.advanceTimersByTime(900);
    process.nextTick(() => {
      expect(wrapper.state('usableSuggestions')).toEqual(['John (Dough)']);
      expect(errorListener).toBeCalledTimes(0);
      expect(symbolListener).toBeCalledTimes(0);
      done();
    });
  });

  // TODO fix this test?
  // It needs to be selected from a list because the same
  // company can trade under multiple symbols
  // jest is not acknowledging the second time advance
  it.skip('calls back when a stock is selected', async (done) => {
    const input = await wrapper.find('.name-input');
    input.simulate('change', { target: { value: 'John' } });
    jest.advanceTimersByTime(900);
    input.simulate('change', { target: { value: 'John (Dough)' } });
    jest.advanceTimersByTime(900);
    process.nextTick(() => {
      expect(symbolListener).toBeCalledWith('Dough', 'John');
      done();
    });
  });

  it('specifies an error on fetch error with a message property', async (done) => {
    fetch.mockReject(new Error('my error message'));
    const input = await wrapper.find('.name-input');
    input.simulate('change', { target: { value: 'john' } });
    jest.advanceTimersByTime(900);
    process.nextTick(() => {
      expect(errorListener).toBeCalledWith(
        'Error fetching symbol from server: my error message'
      );
      expect(symbolListener).toBeCalledTimes(0);
      done();
    });
  });

  it('specifies an error on fetch error without a message property', async (done) => {
    fetch.mockReject('my error message');
    const input = await wrapper.find('.name-input');
    input.simulate('change', { target: { value: 'john' } });
    jest.advanceTimersByTime(900);
    process.nextTick(() => {
      expect(errorListener).toBeCalledWith(
        'Error fetching symbol from server: no message given'
      );
      expect(symbolListener).toBeCalledTimes(0);
      done();
    });
  });
});
