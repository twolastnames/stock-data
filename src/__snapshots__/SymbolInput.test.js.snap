// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders blank state 1`] = `
<div
  className="company-searcher"
>
  <datalist
    id="suggestions"
  />
  <label
    className="name-label"
  >
    Company Name
  </label>
  <input
    autoComplete="off"
    className="name-input"
    list="suggestions"
    onChange={[Function]}
    type="text"
  />
</div>
`;
